File: app/src/lib/watermark/minimap.ts
/**
 * Offline vector mini-map (§5.4 default path).
 *
 * Draws the matched ward (or L&O jurisdiction) outline plus neighbouring
 * boundaries and a pin, from the bundled polygon data, onto a small
 * canvas. Zero network, zero API cost, and deliberately does NOT mimic
 * Google branding — this is our own rendering of our own boundary data.
 *
 * The online Google Static Maps upgrade (backfill path) replaces this
 * canvas with a fetched image and only then shows Google attribution.
 */
import type { Feature, Position } from "geojson";
import { loadGeodataFor } from "../geo/geodata";
import type { LookupResult } from "../geo/lookup";

// 512 so the thumb stays crisp when the card stretches it vertically —
// at 256 the upscale was visibly soft on a full-resolution photo
export const MINIMAP_SIZE = 512;

/**
 * Light paper map with a lit ward.
 *
 * The first version was dark slate with a flat cyan outline and a rose
 * dot — stock dashboard colours that made a real boundary dataset look
 * like filler. This reads as a printed map instead: warm paper, parcels
 * in soft ink, and the ward you are standing in genuinely lit, its glow
 * painted as a blurred pass under a crisp edge rather than a translucent
 * fill pretending to be one.
 */
const INK = {
  paper0: "#fbfcfe",
  paper1: "#e9eff6",
  neighborLine: "rgba(96, 120, 150, 0.32)",
  neighborFill: "rgba(255, 255, 255, 0.55)",
  glow: "rgba(14, 165, 233, 0.55)",
  focusLine: "#0369a1",
  focusA: "rgba(56, 189, 248, 0.42)",
  focusB: "rgba(125, 211, 252, 0.10)",
  pin: "#e11d48",
  pinGlow: "rgba(225, 29, 72, 0.45)",
};

function eachRing(f: Feature, cb: (ring: Position[]) => void): void {
  const g = f.geometry;
  if (g.type === "Polygon") for (const r of g.coordinates) cb(r);
  else if (g.type === "MultiPolygon")
    for (const poly of g.coordinates) for (const r of poly) cb(r);
}

let cacheKey = "";
let cacheCanvas: HTMLCanvasElement | null = null;

/**
 * Render the offline mini-map for a location. Cached by focus feature +
 * ~30 m position cell, so live-preview calls are nearly free.
 */
export async function renderMiniMap(
  lat: number,
  lng: number,
  lookupResult: LookupResult | null
): Promise<HTMLCanvasElement | null> {
  const focus =
    lookupResult?.wardFeature ?? lookupResult?.loFeature ?? null;

  const key = `${focus ? JSON.stringify(focus.bbox) : "none"}|${lat.toFixed(4)},${lng.toFixed(4)}`;
  if (key === cacheKey && cacheCanvas) return cacheCanvas;

  const pack = await loadGeodataFor(lat, lng).catch(() => null);
  if (!pack) return null;

  const size = MINIMAP_SIZE;
  const canvas = document.createElement("canvas");
  canvas.width = size;
  canvas.height = size;
  const ctx = canvas.getContext("2d");
  if (!ctx) return null;

  // View window: the focus polygon's bbox padded 25%, or ~600 m around
  // the point when nothing matched.
  let [minX, minY, maxX, maxY] = focus?.bbox ?? [
    lng - 0.006,
    lat - 0.006,
    lng + 0.006,
    lat + 0.006,
  ];
  // keep the pin inside the window
  minX = Math.min(minX, lng);
  maxX = Math.max(maxX, lng);
  minY = Math.min(minY, lat);
  maxY = Math.max(maxY, lat);
  const padX = (maxX - minX) * 0.25 || 0.002;
  const padY = (maxY - minY) * 0.25 || 0.002;
  minX -= padX; maxX += padX; minY -= padY; maxY += padY;

  // Square aspect: expand the shorter axis (lat degrees ≈ lng degrees
  // near the equator is close enough at Chennai's latitude for a thumb).
  const spanX = maxX - minX;
  const spanY = maxY - minY;
  if (spanX > spanY) {
    const grow = (spanX - spanY) / 2;
    minY -= grow; maxY += grow;
  } else {
    const grow = (spanY - spanX) / 2;
    minX -= grow; maxX += grow;
  }

  const px = (x: number) => ((x - minX) / (maxX - minX)) * size;
  const py = (y: number) => size - ((y - minY) / (maxY - minY)) * size;

  // unit scale so every stroke and radius tracks MINIMAP_SIZE
  const u = size / 256;

  // paper: soft radial lift towards the middle, so the tile has depth
  // without a border or a drop shadow doing the work
  const paper = ctx.createRadialGradient(
    size / 2, size * 0.42, size * 0.08,
    size / 2, size / 2, size * 0.78
  );
  paper.addColorStop(0, INK.paper0);
  paper.addColorStop(1, INK.paper1);
  ctx.fillStyle = paper;
  ctx.fillRect(0, 0, size, size);

  const drawFeature = (f: Feature, stroke: string, fill?: string): void => {
    ctx.beginPath();
    eachRing(f, (ring) => {
      ring.forEach(([x, y], i) => {
        if (i === 0) ctx.moveTo(px(x), py(y));
        else ctx.lineTo(px(x), py(y));
      });
      ctx.closePath();
    });
    if (fill) {
      ctx.fillStyle = fill;
      ctx.fill();
    }
    ctx.strokeStyle = stroke;
    ctx.lineWidth = (fill ? 2 : 1) * u;
    ctx.stroke();
  };

  // Neighbour context: ward boundaries whose bbox overlaps the window.
  const layer = lookupResult?.wardFeature ? pack.layers.ulb : pack.layers.lo;
  let drawn = 0;
  for (const f of layer.features as Feature[]) {
    if (!f.bbox) continue;
    if (f.bbox[0] > maxX || f.bbox[2] < minX || f.bbox[1] > maxY || f.bbox[3] < minY) continue;
    if (f === focus) continue;
    drawFeature(f, INK.neighborLine, INK.neighborFill);
    if (++drawn > 40) break; // thumbnails don't need more context than this
  }

  if (focus) {
    // the lit ward: a blurred pass lays the halo down, then a clean edge
    // is drawn over it. Faking the glow with one translucent stroke is
    // what made the old tile look flat.
    ctx.save();
    ctx.shadowColor = INK.glow;
    ctx.shadowBlur = 16 * u;
    ctx.lineJoin = "round";
    const grad = ctx.createLinearGradient(0, 0, 0, size);
    grad.addColorStop(0, INK.focusA);
    grad.addColorStop(1, INK.focusB);
    drawFeature(focus, INK.focusLine, grad as unknown as string);
    ctx.shadowBlur = 8 * u;
    drawFeature(focus, INK.focusLine);
    ctx.restore();
    drawFeature(focus, INK.focusLine);
  }

  // Pin: halo, white collar, core — reads at thumbnail size where a flat
  // dot just disappears into the boundary lines
  const cx = px(lng);
  const cy = py(lat);
  const halo = ctx.createRadialGradient(cx, cy, 0, cx, cy, 15 * u);
  halo.addColorStop(0, INK.pinGlow);
  halo.addColorStop(1, "rgba(225, 29, 72, 0)");
  ctx.fillStyle = halo;
  ctx.beginPath();
  ctx.arc(cx, cy, 15 * u, 0, Math.PI * 2);
  ctx.fill();

  ctx.beginPath();
  ctx.arc(cx, cy, 7 * u, 0, Math.PI * 2);
  ctx.fillStyle = "#ffffff";
  ctx.fill();

  ctx.beginPath();
  ctx.arc(cx, cy, 4.6 * u, 0, Math.PI * 2);
  ctx.fillStyle = INK.pin;
  ctx.fill();

  cacheKey = key;
  cacheCanvas = canvas;
  return canvas;
}

File: native-android/app/src/main/java/city/reclaimchennai/gpscam/wm/MiniMap.kt
package city.reclaimchennai.gpscam.wm

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import city.reclaimchennai.gpscam.geo.GeoFeature
import city.reclaimchennai.gpscam.geo.GeoPack
import city.reclaimchennai.gpscam.geo.LookupResult
import kotlin.math.max
import kotlin.math.min

/**
 * Offline vector mini-map: matched ward outline + neighbour boundaries +
 * a pin, drawn from the pack polygons. Port of the web renderer, same
 * palette, no third-party tiles.
 */
object MiniMap {
    private const val SIZE = 256

    fun render(pack: GeoPack, result: LookupResult?, lat: Double, lng: Double): Bitmap {
        val focus = result?.wardFeature ?: result?.loFeature
        val bmp = Bitmap.createBitmap(SIZE, SIZE, Bitmap.Config.ARGB_8888)
        val c = Canvas(bmp)
        c.drawColor(Color.parseColor("#17212b"))

        var minX: Double; var minY: Double; var maxX: Double; var maxY: Double
        val fb = focus?.bbox
        if (fb != null) {
            minX = fb[0]; minY = fb[1]; maxX = fb[2]; maxY = fb[3]
        } else {
            minX = lng - 0.006; minY = lat - 0.006
            maxX = lng + 0.006; maxY = lat + 0.006
        }
        minX = min(minX, lng); maxX = max(maxX, lng)
        minY = min(minY, lat); maxY = max(maxY, lat)
        var padX = (maxX - minX) * 0.25
        if (padX == 0.0) padX = 0.002
        var padY = (maxY - minY) * 0.25
        if (padY == 0.0) padY = 0.002
        minX -= padX; maxX += padX; minY -= padY; maxY += padY
        val spanX = maxX - minX; val spanY = maxY - minY
        if (spanX > spanY) { val g = (spanX - spanY) / 2; minY -= g; maxY += g }
        else { val g = (spanY - spanX) / 2; minX -= g; maxX += g }

        fun px(x: Double) = ((x - minX) / (maxX - minX) * SIZE).toFloat()
        fun py(y: Double) = (SIZE - (y - minY) / (maxY - minY) * SIZE).toFloat()

        val stroke = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE }
        val fill = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }

        fun draw(f: GeoFeature, color: Int, fillColor: Int?) {
            val path = Path()
            for (rings in f.polygons) for (ring in rings) {
                for (i in 0 until ring.size / 2) {
                    val x = px(ring[i * 2]); val y = py(ring[i * 2 + 1])
                    if (i == 0) path.moveTo(x, y) else path.lineTo(x, y)
                }
                path.close()
            }
            if (fillColor != null) {
                fill.color = fillColor
                c.drawPath(path, fill)
            }
            stroke.color = color
            stroke.strokeWidth = if (fillColor != null) 2f else 1f
            c.drawPath(path, stroke)
        }

        val layer = if (result?.wardFeature != null) pack.ulb else pack.lo
        var drawn = 0
        for (f in layer.features) {
            val b = f.bbox ?: continue
            if (b[0] > maxX || b[2] < minX || b[1] > maxY || b[3] < minY) continue
            if (f === focus) continue
            draw(f, Color.argb(115, 148, 163, 184), null)
            if (++drawn > 40) break
        }
        focus?.let { draw(it, Color.parseColor("#38bdf8"), Color.argb(36, 56, 189, 248)) }

        val cx = px(lng); val cy = py(lat)
        fill.color = Color.parseColor("#f43f5e")
        c.drawCircle(cx, cy, 7f, fill)
        stroke.color = Color.WHITE
        stroke.strokeWidth = 2.5f
        c.drawCircle(cx, cy, 7f, stroke)
        return bmp
    }
}
